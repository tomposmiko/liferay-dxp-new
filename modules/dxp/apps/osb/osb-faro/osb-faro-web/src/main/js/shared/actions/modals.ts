import {Modal} from 'shared/types';

const actionTypes = Modal.actionTypes;
const modalTypes = Modal.modalTypes;

export {actionTypes, modalTypes};

export const close = () => ({
	type: actionTypes.CLOSE_MODAL
});

export const closeAll = () => ({
	type: actionTypes.CLOSE_ALL_MODALS
});

export const open = (
	type,
	props: {[key: string]: any} = {},
	options: {closeOnBlur?: boolean} = {}
) => {
	const {closeOnBlur = true} = options;

	return {
		payload: {
			closeOnBlur,
			props,
			type
		},
		type: actionTypes.OPEN_MODAL
	};
};
