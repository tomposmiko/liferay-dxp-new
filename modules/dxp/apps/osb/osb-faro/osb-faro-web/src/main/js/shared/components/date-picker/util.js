import moment from 'moment';
import {conformsTo, isNil} from 'lodash';

/**
 * Represents a time between two dates.
 * @typedef Range
 * @property {Moment} start - The start of the range.
 * @property {Moment} end - The end of the range.
 */

export function isAboveMaxRange(date, maxRange) {
	return (
		isRange(date) &&
		date.start &&
		date.end &&
		maxRange &&
		moment
			.duration({
				from: date.start,
				to: date.end
			})
			.asDays() > maxRange
	);
}

/**
 * A validation function that determines if a value is a valid
 * moment instance, allowing for null.
 * @param {Object} value - The value to test.
 * @returns {boolean}
 */
function isMomentOrNil(value) {
	return isNil(value) || moment.isMoment(value);
}

/**
 * A validation function that determines if the value is a valid
 * moment or range.
 * @param {Object} value - The value to test.
 * @returns {boolean}
 */
export function isDateOrRange(value) {
	return isNil(value) || moment.isMoment(value) || isRange(value);
}

/**
 * Determines if a date is in a {@link Range}.
 * @param {Range|Moment} dateOrRange - The range to check in.
 * @param {Moment} date - The date to check for.
 * @returns {boolean}
 */
export function isInRange(dateOrRange, date) {
	if (moment.isMoment(dateOrRange)) {
		return false;
	}

	const {end, start} = dateOrRange;

	return start && date.isAfter(start) && end && date.isBefore(end);
}

/**
 * A validation function that determines if a value is a valid {@link Range}.
 * @param {Object} value - The value to test.
 * @returns {boolean}
 */
export function isRange(value) {
	return conformsTo(value, {
		end: isMomentOrNil,
		start: isMomentOrNil
	});
}

/**
 * Adds a date to an existing {@link Range}.
 *
 * This function will try to be smart and first update the
 * start or end if either are null. If both exist, then it will return
 * a new range with the start set to the new date, and the end reset
 * to null.
 * @param {Range} range - The range to update.
 * @param {Moment} date - The date to add to the range.
 * @returns {Range}
 */
export function updateRange(range, date) {
	if (!range.start) {
		return {
			...range,
			start: date
		};
	}

	if (range.start && !range.end) {
		if (date.isSameOrAfter(range.start)) {
			return {
				...range,
				end: date
			};
		} else {
			return {
				end: range.start,
				start: date
			};
		}
	}

	if (date.isSame(range.start) && date.isSame(range.end)) {
		return range;
	}

	return {
		end: null,
		start: date
	};
}
