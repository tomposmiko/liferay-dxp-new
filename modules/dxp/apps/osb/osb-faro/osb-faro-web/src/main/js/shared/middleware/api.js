import sendRequest from 'shared/util/request';
import {get, isNil} from 'lodash';

export const CALL_API = 'CALL_API';

export function toAction(type, ...objs) {
	const action = Object.assign(...objs, {type});

	delete action.meta[CALL_API];

	return action;
}

export default () => next => action => {
	const request = get(action, ['meta', CALL_API]);

	if (isNil(request)) {
		return next(action);
	}

	const {data, requestFn, types} = request;

	const [requestType, successType, failureType] = types;

	next(toAction(requestType, action));

	const retVal = requestFn ? requestFn(data) : sendRequest(request);

	return retVal.then(
		payload => {
			next(
				toAction(successType, action, {
					meta: {
						...action.meta,
						schema: request.schema
					},
					payload
				})
			);

			return {payload};
		},
		error => {
			next(
				toAction(failureType, action, {
					error: true
				})
			);

			throw error;
		}
	);
};
