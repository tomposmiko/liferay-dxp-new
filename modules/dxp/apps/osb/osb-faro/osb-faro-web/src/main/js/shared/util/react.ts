import {isEqual} from 'lodash';

export const getDisplayName = WrappedComponent =>
	WrappedComponent.displayName || WrappedComponent.name || 'Component';

type HasChanges = <T>(
	prev: T | object,
	next: T | object,
	...keys: string[]
) => boolean;

/**
 * Compare previous state or props object by provided keys to detect changes.
 */
export const hasChanges: HasChanges = (prev = {}, next = {}, ...keys) => {
	for (const key of keys) {
		if (key in next) {
			const newVal = next[key];

			const prevVal = prev[key];

			if (!isEqual(newVal, prevVal)) {
				return true;
			}
		}
	}

	return false;
};
