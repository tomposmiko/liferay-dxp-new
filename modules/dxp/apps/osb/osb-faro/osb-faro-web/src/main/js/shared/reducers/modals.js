import {actionTypes} from '../actions/modals';
import {createReducer} from 'redux-toolbox';
import {List, Map} from 'immutable';

export default createReducer(new List(), {
	[actionTypes.CLOSE_ALL_MODALS](state) {
		return state.clear();
	},

	[actionTypes.CLOSE_MODAL](state) {
		return state.pop();
	},

	[actionTypes.OPEN_MODAL](state, action) {
		const {closeOnBlur, props, type} = action.payload;

		return state.push(
			new Map({
				closeOnBlur,
				props: new Map(props),
				type
			})
		);
	}
});
