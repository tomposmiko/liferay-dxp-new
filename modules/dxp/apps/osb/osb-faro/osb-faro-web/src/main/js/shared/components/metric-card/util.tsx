import MetricValue from './MetricValue';
import React, {Fragment} from 'react';
import Trend from 'shared/components/Trend';
import {CHART_COLOR_NAMES} from 'shared/components/Chart';
import {get, last} from 'lodash';
import {
	getAxisFormatter,
	getDataFormatter,
	getIntervals,
	getMetricFormatter
} from 'shared/util/charts';
import {getIcon, getStatsColor} from 'shared/util/metrics';
import {INTERVAL_KEY_MAP} from 'shared/util/time';
import {MetricType} from './metrics';
import {toRounded} from 'shared/util/numbers';
import {toUnix} from 'shared/util/date';

const CHART_DATA_ID_1 = 'data_1';
const CHART_DATA_ID_2 = 'data_2';
const CHART_DATA_PREVIOUS = 'data_previous';
const METRIC_TOOLTIP_LABEL_MAP = {
	bounceRateMetric: Liferay.Language.get('avg-bounce')
};

const {
	martell: CHART_GREEN,
	martellL2: CHART_GREEN_L2,
	mormont: CHART_ORANGE,
	stark: CHART_BLUE,
	starkL2: CHART_BLUE_L2
} = CHART_COLOR_NAMES;

export const Icons = {
	negative: 'caret-bottom-l',
	neutral: undefined,
	positive: 'caret-top-l'
};

const PREVIOUS_PERIOD_VISITORS_COLOR = '#393A4A';

type TBuildTabs = (props: {
	activeItemIndex: number;
	items: {
		content: {
			details: {
				color: string;
				icon: string;
				label: string;
			};
			title: string;
			type: MetricType;
			value: string;
		};
	}[];
	onActiveItemIndexChange: (index: number) => void;
}) => {
	onClick: () => void | ((index: number) => void);
	secondaryInfo: React.ReactElement;
}[];

export const buildTabs: TBuildTabs = ({
	activeItemIndex,
	items,
	onActiveItemIndexChange
}) =>
	items.map(({content}, index) => {
		const {details, title, type, value} = content;
		const {color, icon, label} = details;

		return {
			onClick: () => {
				if (activeItemIndex !== index && !onActiveItemIndexChange)
					return () => {};

				return onActiveItemIndexChange(index);
			},
			secondaryInfo: (
				<span>
					<span className='primary-content'>
						<MetricValue type={type} value={value} />
					</span>

					{label && <Trend color={color} icon={icon} label={label} />}
				</span>
			),
			tabId: index,
			title
		};
	});

export const getMetricName = (activeItemIndex, metrics) =>
	metrics.map(({name}) => name)[activeItemIndex];

export const getActiveItem = (retVal, compareToPrevious) => {
	if (!retVal) {
		return {
			chartData: [],
			intervals: [],
			timeline: []
		};
	}

	const chartData = retVal.data.slice(0, -1);
	const timeline = last(retVal.data);

	if (!compareToPrevious && retVal.asymmetricComparison) {
		retVal = {
			...retVal,
			chartData: chartData.map(dataSet => ({
				...dataSet,
				data: dataSet.data.slice(1)
			})),
			intervals: retVal.intervals.slice(1),

			// @ts-ignore

			timeline: {data: timeline.data.slice(1), id: timeline.id}
		};

		if (retVal.compositeData) {
			const compositeDataKeys = Object.keys(retVal.compositeData);

			retVal = {
				...retVal,
				compositeData: compositeDataKeys.reduce((acc, val) => {
					acc = {
						...acc,
						[val]: retVal.compositeData[val].slice(1)
					};

					return acc;
				}, {})
			};
		}
	} else if (compareToPrevious && retVal.asymmetricComparison) {
		retVal = {
			...retVal,
			chartData: chartData.map(dataSet => ({
				...dataSet,
				data:
					dataSet.id !== 'data_previous'
						? [null, ...dataSet.data.slice(1)]
						: dataSet.data
			})),
			timeline
		};
	} else {
		retVal = {
			...retVal,
			chartData,
			timeline
		};
	}

	return retVal;
};

export const getPreviousValueFromCompositeData = (
	compositeData,
	dataName,
	dateKey
) => {
	const data = get(compositeData, dataName);

	if (data) {
		return data.find(val => toUnix(val.key) === dateKey).previousValue;
	}
};

export const getRegexType = (type: MetricType): RegExp => {
	if (type === MetricType.Ratings) {
		return /([/][0-9]+)/g;
	} else {
		return /([a-zA-Z%])+/g;
	}
};

export const formatValue = (
	value: string,
	regex: RegExp
): React.ReactElement[] => {
	const items = value.split(' ');

	return items.map((item, i) => {
		const [value, unit] = item.split(regex);

		return (
			<Fragment key={i}>
				{value}

				<span className='metric-value-letter'>{unit}</span>
			</Fragment>
		);
	});
};

export const getMetricCardTabsData = (result, metrics) =>
	metrics.map(({name, title, type}) => {
		const metricFormatter = getMetricFormatter(type);

		return {
			content: {
				details: {
					color: getStatsColor(
						result[name].trend.trendClassification
					),
					icon: getIcon(result[name].trend.percentage),
					label: `${toRounded(
						Math.abs(result[name].trend.percentage)
					)}%`
				},
				name,
				title,
				type,
				value: metricFormatter(result[name].value)
			}
		};
	});

export const getMetricsData = (
	result,
	metrics,
	rangeSelectors = {},
	chartDataMapFn = getMetricsChartData,
	interval = INTERVAL_KEY_MAP.day
) =>
	metrics.map(({compositeMetrics, name, title, tooltipTitle, type}) =>
		getMetricData({
			chartDataMapFn,
			compositeMetrics,
			interval,
			name,
			rangeSelectors,
			result,
			title,
			tooltipTitle,
			type
		})
	);

export const getMetricsChartData = ({
	histogram,
	name,
	title,
	tooltipTitle,
	type
}) => [
	{
		color: CHART_BLUE,
		data: getDataFormatter(type)(histogram.map(({value}) => value)),
		id: CHART_DATA_ID_1,
		name: tooltipTitle || METRIC_TOOLTIP_LABEL_MAP[name] || title,
		tooltipTitle
	},
	{
		color: CHART_BLUE_L2,
		data: getDataFormatter(type)(
			histogram.map(({previousValue}) => previousValue)
		),
		id: CHART_DATA_PREVIOUS,
		name: Liferay.Language.get('previous-period')
	},
	{
		data: histogram.map(({key}) => key),
		id: 'x'
	}
];

export const convertHistogramKeysToDate = ({
	key,
	previousValueKey,
	valueKey,
	...otherParams
}) => ({
	key: toUnix(key),
	previousValueKey: previousValueKey.split('/').map(toUnix),
	valueKey: valueKey.split('/').map(toUnix),
	...otherParams
});

export const getMetricData = ({
	chartDataMapFn = getMetricsChartData,
	compositeMetrics,
	interval = INTERVAL_KEY_MAP.day,
	name,
	rangeSelectors,
	result,
	title,
	tooltipTitle,
	type
}) => {
	const metricFormatter = getMetricFormatter(type);

	const histogram = result[name].histogram.metrics.map(
		convertHistogramKeysToDate
	);

	const compositeData = compositeMetrics
		? {
				compositeContent: compositeMetrics.reduce(
					(
						acc,
						{
							name: compositeMetricName,
							title: compositeMetricTitle,
							type: compositeMetricType
						}
					) => {
						acc[compositeMetricName] = {
							details: {
								color: getStatsColor(
									result[compositeMetricName].trend
										.trendClassification
								),
								icon: getIcon(
									result[compositeMetricName].trend.percentage
								),
								label: `${toRounded(
									Math.abs(
										result[compositeMetricName].trend
											.percentage
									)
								)}%`
							},
							name: compositeMetricName,
							title: compositeMetricTitle,
							type: compositeMetricType,
							value: metricFormatter(
								result[compositeMetricName].value
							)
						};

						return acc;
					},
					{}
				),
				compositeData: compositeMetrics.reduce(
					(acc, {name: compositeMetricName}) => {
						acc[compositeMetricName] =
							result[compositeMetricName].histogram.metrics;

						return acc;
					},
					{}
				)
		  }
		: {};

	const dateKeysIMap = new Map(
		histogram.map(({key, valueKey}) => [key, valueKey])
	);

	return {
		...compositeData,
		asymmetricComparison: result[name].histogram.asymmetricComparison,
		content: {
			details: {
				color: getStatsColor(result[name].trend.trendClassification),
				icon: getIcon(result[name].trend.percentage),
				label: `${toRounded(Math.abs(result[name].trend.percentage))}%`
			},
			name,
			title,
			type,
			value: metricFormatter(result[name].value)
		},
		data: chartDataMapFn({
			...compositeData,
			histogram,
			name,
			title,
			tooltipTitle,
			type
		}),
		dateKeysIMap,
		format: getAxisFormatter(type),
		intervals: getIntervals(
			rangeSelectors.rangeKey,
			histogram.map(({key}) => key),
			interval,
			dateKeysIMap
		),
		prevDateKeysIMap: new Map(
			histogram.map(({key, previousValueKey}) => [key, previousValueKey])
		)
	};
};

export const getSiteMetricsChartData = ({
	compositeData,
	histogram,
	name,
	title,
	tooltipTitle,
	type
}) =>
	name === 'visitorsMetric'
		? [
				{
					color: CHART_BLUE,
					data: getDataFormatter(type)(
						compositeData.knownVisitorsMetric.map(
							({value}) => value
						)
					),
					dataName: 'knownVisitorsMetric',
					id: CHART_DATA_ID_1,
					name: Liferay.Language.get('known-visitors'),
					tooltipTitle: Liferay.Language.get('known'),
					type: 'bar'
				},
				{
					color: CHART_ORANGE,
					data: getDataFormatter(type)(
						compositeData.anonymousVisitorsMetric.map(
							({value}) => value
						)
					),
					dataName: 'anonymousVisitorsMetric',
					id: CHART_DATA_ID_2,
					name: Liferay.Language.get('anonymous-visitors'),
					tooltipTitle: Liferay.Language.get('anonymous'),
					type: 'bar'
				},
				{
					color: PREVIOUS_PERIOD_VISITORS_COLOR,
					data: getDataFormatter(type)(
						histogram.map(({previousValue}) => previousValue)
					),
					id: CHART_DATA_PREVIOUS,
					name: Liferay.Language.get('previous-period'),
					type: 'line'
				},
				{
					data: histogram.map(({key}) => key),
					id: 'x'
				}
		  ]
		: getMetricsChartData({histogram, name, title, tooltipTitle, type}).map(
				data =>
					[CHART_DATA_ID_1, CHART_DATA_PREVIOUS].includes(data.id)
						? {
								...data,
								color:
									data.id === CHART_DATA_PREVIOUS
										? CHART_GREEN_L2
										: CHART_GREEN
						  }
						: data
		  );
