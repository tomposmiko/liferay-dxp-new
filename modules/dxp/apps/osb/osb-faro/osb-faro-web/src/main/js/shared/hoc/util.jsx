import ErrorDisplay from 'shared/components/ErrorDisplay';
import ErrorPage from 'shared/pages/ErrorPage';
import LoadingPage from 'shared/pages/Loading';
import NoResultsDisplay from 'shared/components/NoResultsDisplay';
import React from 'react';
import Spinner from 'shared/components/Spinner';
import {compose} from 'redux';
import {get, omit} from 'lodash';
import {sub} from 'shared/util/lang';
import {toRoute} from 'shared/util/router';

/**
 * HOC for ErrorDisplay.
 * @param {Object} - Options object to pass as props to ErrorDisplay
 * @returns {Function} Returns the ErrorDisplay or Wrapped Component.
 */
export const withError =
	(options = {}) =>
	Component =>
	({error, errorProps = {}, pageDisplay = true, refetch, ...otherProps}) => {
		const otherOptions = omit(options, 'page');

		if (error) {
			return get(options, 'page', pageDisplay) ? (
				<ErrorPage {...errorProps} {...otherOptions} />
			) : (
				<ErrorDisplay
					onReload={refetch}
					spacer
					{...errorProps}
					{...otherOptions}
				/>
			);
		}

		return <Component refetch={refetch} {...otherProps} />;
	};

/**
 * HOC for NoResultsDisplay.
 * @param {Object} - Options object to pass as props to NoResultsDisplay.
 * @returns {Function} Returns the NoResultsDisplay or WrappedComponent.
 */
export const withEmpty =
	(options = {}) =>
	Component =>
	({
		data,
		error,
		loading,
		noResultsRenderer,
		noResultsRendererProps,
		total,
		...props
	}) => {
		if (((data && data.total === 0) || total === 0) && !loading && !error) {
			if (noResultsRenderer) {
				const NoResults = noResultsRenderer;

				return <NoResults />;
			}

			return (
				<NoResultsDisplay {...options} {...noResultsRendererProps} />
			);
		}

		return (
			<Component
				data={data}
				error={error}
				loading={loading}
				total={total}
				{...props}
			/>
		);
	};

/**
 * HOC for Loading display.
 * @param {Object} - Options object to pass as props to Loading component.
 * @returns {Function} Returns the Loading or WrappedComponent.
 */
export const withLoading =
	(options = {}) =>
	Component =>
	({
		alignCenter = false,
		className,
		data,
		fadeIn = true,
		inline = false,
		loading,
		pageDisplay = true,
		...otherProps
	}) => {
		if (loading) {
			return get(options, 'page', pageDisplay) ? (
				<LoadingPage
					className={className}
					fadeIn={get(options, 'fadeIn', fadeIn)}
					key='LOADING'
				/>
			) : (
				<Spinner
					alignCenter={get(options, 'alignCenter', alignCenter)}
					className={className}
					inline={get(options, 'inline', inline)}
					key='SPINNER'
					spacer={!get(options, 'inline', inline)}
				/>
			);
		}

		return <Component className={className} data={data} {...otherProps} />;
	};

export const withNull =
	(key, errorProps = {}) =>
	Component =>
	props => {
		const {entityType = Liferay.Language.get('page'), linkRoute} =
			errorProps;

		if (key && !props[key]) {
			return (
				<ErrorPage
					{...props}
					href={toRoute(linkRoute, props.router.params)}
					linkLabel={sub(Liferay.Language.get('go-to-x'), [
						entityType
					])}
					message={sub(
						Liferay.Language.get(
							'the-x-you-are-looking-for-does-not-exist'
						),
						[entityType.toLowerCase()]
					)}
					subtitle={sub(Liferay.Language.get('x-not-found'), [
						entityType
					])}
					title={Liferay.Language.get('404')}
				/>
			);
		}

		return <Component {...props} />;
	};

/**
 * HOC for displaying results.
 */
export const SafeResults = compose(
	withLoading(),
	withError()
)(({children, data}) => children(data));

export const WrapSafeResults = compose(
	withLoading(),
	withError()
)(({children}) => children);
