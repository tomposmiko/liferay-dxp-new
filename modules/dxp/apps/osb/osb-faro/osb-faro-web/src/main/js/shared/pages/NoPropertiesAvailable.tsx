import * as API from 'shared/api';
import BasePage from 'shared/components/base-page';
import ClayLink from '@clayui/link';
import Constants, {Sizes} from 'shared/util/constants';
import NoResultsDisplay from 'shared/components/NoResultsDisplay';
import React from 'react';
import {close, modalTypes, open} from 'shared/actions/modals';
import {compose} from 'redux';
import {connect} from 'react-redux';
import {Routes, toRoute} from 'shared/util/router';
import {setBackURL} from 'shared/actions/settings';
import {User} from 'shared/util/records';
import {withRequest} from 'shared/hoc';

const {
	pagination: {cur: defaultPage}
} = Constants;

interface INoPropertiesAvailableProps
	extends React.HTMLAttributes<HTMLDivElement> {
	close: () => void;
	currentUser: User;
	dataSources: boolean;
	groupId: string;
	open: (modalType: string, config: object) => void;
	setBackURL: (url: string) => void;
}

const NoPropertiesAvailable: React.FC<INoPropertiesAvailableProps> = ({
	close,
	currentUser,
	dataSources,
	groupId,
	open,
	setBackURL
}) => {
	const admin = currentUser.isAdmin();

	const description = admin
		? Liferay.Language.get(
				'to-get-your-workspace-set-up-you-will-need-to-create-a-property-and-sync-your-dxp-sites'
		  )
		: Liferay.Language.get(
				'you-have-not-been-added-to-any-properties.-please-contact-your-analytics-cloud-administrator'
		  );

	const title = admin
		? Liferay.Language.get('first-connect-your-dxp-sites')
		: Liferay.Language.get('no-properties-found');

	return (
		<BasePage
			className='no-properties-available-root'
			documentTitle={Liferay.Language.get('no-properties-available')}
		>
			<BasePage.Header breadcrumbs={[]} groupId={groupId}>
				<BasePage.Header.TitleSection
					title={Liferay.Language.get('no-properties-available')}
				/>
			</BasePage.Header>

			<BasePage.Body>
				<NoResultsDisplay
					description={
						<>
							<p>
								{dataSources
									? description
									: Liferay.Language.get(
											'first-complete-the-onboarding-to-get-your-workspace-set-up'
									  )}
							</p>

							{admin && (
								<ClayLink
									button
									className='button-root'
									displayType='primary'
									href={
										dataSources
											? toRoute(
													Routes.SETTINGS_CHANNELS,
													{
														groupId
													}
											  )
											: null
									}
									onClick={
										dataSources
											? () =>
													setBackURL(
														toRoute(
															Routes.WORKSPACE_WITH_ID,
															{
																groupId
															}
														)
													)
											: () =>
													open(
														modalTypes.ONBOARDING_MODAL,
														{
															groupId,
															onClose: close
														}
													)
									}
								>
									{dataSources
										? Liferay.Language.get(
												'create-property'
										  )
										: Liferay.Language.get('start')}
								</ClayLink>
							)}
						</>
					}
					displayCard
					icon={{
						border: false,
						size: Sizes.XXXLarge,
						symbol: 'ac-satellite'
					}}
					spacer
					title={
						dataSources
							? title
							: Liferay.Language.get('you-are-almost-there')
					}
				/>
			</BasePage.Body>
		</BasePage>
	);
};

export default compose<any>(
	withRequest(
		({groupId}) =>
			API.dataSource.search({
				delta: 1,
				groupId,
				page: defaultPage,
				query: ''
			}),
		({total}) => ({
			dataSources: !!total
		})
	),
	connect(null, {close, open, setBackURL})
)(NoPropertiesAvailable);
