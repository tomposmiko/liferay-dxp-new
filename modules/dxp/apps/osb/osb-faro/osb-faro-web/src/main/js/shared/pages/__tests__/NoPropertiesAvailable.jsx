jest.mock('shared/actions/modals', () => ({
	actionTypes: {},
	close: jest.fn(),
	modalTypes: {},
	open: jest.fn(() => ({meta: {}, payload: {}, type: 'open'}))
}));

import * as API from 'shared/api';
import mockStore from 'test/mock-store';
import NoPropertiesAvailable from '../NoPropertiesAvailable';
import Promise from 'metal-promise';
import React from 'react';
import {cleanup, fireEvent, render} from '@testing-library/react';
import {getImmutableMock, mockMemberUser, mockUser} from 'test/data';
import {open} from 'shared/actions/modals';
import {Provider} from 'react-redux';
import {StaticRouter} from 'react-router';
import {User} from 'shared/util/records';

jest.unmock('react-dom');

const DefaultComponent = props => (
	<Provider store={mockStore()}>
		<StaticRouter>
			<NoPropertiesAvailable
				currentUser={getImmutableMock(User, mockUser)}
				groupId='123123'
				{...props}
			/>
		</StaticRouter>
	</Provider>
);

describe('NoPropertiesAvailable', () => {
	afterEach(cleanup);

	it('should render', () => {
		API.dataSource.search.mockReturnValueOnce(Promise.resolve({total: 0}));

		const {container} = render(<DefaultComponent />);

		jest.runAllTimers();

		expect(container).toMatchSnapshot();
	});

	it('should render w/ Create Property button', () => {
		const {queryByText} = render(<DefaultComponent />);

		jest.runAllTimers();

		expect(queryByText('Create Property')).toBeTruthy();
	});

	it('should not display the "Start" button to trigger onboarding if the user is not an AC admin', () => {
		const {queryByRole} = render(
			<DefaultComponent
				currentUser={getImmutableMock(User, mockMemberUser)}
			/>
		);

		jest.runAllTimers();

		expect(queryByRole('button')).toBeNull();
	});

	it('should call open on "Start" click', () => {
		API.dataSource.search.mockReturnValueOnce(Promise.resolve({total: 0}));

		const {queryByText} = render(<DefaultComponent open={open} />);

		expect(open).not.toBeCalled();

		jest.runAllTimers();

		fireEvent.click(queryByText('Start'));

		expect(open).toBeCalled();
	});
});
