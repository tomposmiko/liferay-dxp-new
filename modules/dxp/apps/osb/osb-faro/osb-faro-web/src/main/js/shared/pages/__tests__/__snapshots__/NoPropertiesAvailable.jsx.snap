// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NoPropertiesAvailable should render 1`] = `
<div>
  <div
    class="index-root no-properties-available-root"
  >
    <div
      class="maintenance-alert-root"
    />
    <header
      class="header-root"
    >
      <div
        class="header-container"
      >
        <div
          class="row-root align-items-center d-flex justify-content-between"
        >
          <ol
            class="breadcrumb breadcrumbs-root"
          />
        </div>
        <div
          class="header-section title-section"
        >
          <span
            class="align-items-center d-flex"
          >
            <h1
              class="title text-truncate"
            >
              <span
                class="text-truncate"
                data-tooltip="false"
                title=""
              >
                No Properties Available
              </span>
            </h1>
          </span>
        </div>
      </div>
    </header>
    <div
      class="body-root page-container"
    >
      <span
        class="children-wrapper"
      >
        <div
          class="no-results-root flex-grow-1 display-card"
        >
          <div
            class="no-results-content spacer"
          >
            <div
              class="no-results-icon"
            >
              <svg
                class="lexicon-icon lexicon-icon-ac-satellite icon-root icon-size-xxxl"
                role="presentation"
              >
                <use
                  xlink:href="#ac-satellite"
                />
              </svg>
            </div>
            <h4
              class="no-results-title"
            >
              You are Almost There
            </h4>
            <div
              class="no-results-description"
            >
              <p>
                First, complete the onboarding to get your workspace set up.
              </p>
              <a
                class="button-root btn btn-primary"
              >
                Start
              </a>
            </div>
          </div>
        </div>
      </span>
    </div>
  </div>
</div>
`;
